peps/pep-0729.rst

```diff
@@ -0,0 +1,414 @@
+the type system less accessible to newcomers. For example:
+
+* How ``@overload`` matching works
+* How ``ParamSpec`` `should work <https://github.com/python/typing/discussions/946>`_ `with methods <https://github.com/microsoft/pyright/issues/3954#issuecomment-1250098464>`_
+* The concept of `recursive aliases <https://github.com/python/typing/issues/182>`_
+* Semantics of `variable initialization <https://mail.python.org/archives/list/typing-sig@python.org/thread/GYVM5KEE6URE6PAH7UTK6324M7GWSFQS/#SY3KPJCAW4UTOOCH3XRJYROSGDEGOTWI>`_
+* `Reachability semantics <https://github.com/python/mypy/issues/15158#issuecomment-1677915108>`__
+  of annotations on ``__exit__``
+* `Symbol visibility <https://mail.python.org/archives/list/typing-sig@python.org/thread/YLJPWECBNPD2K4TRIBRIPISNUZJCRREY/#OX4GLBQOOCMRE5YPZEY3R3XNV6DD7XLW>`_
+* Use of `NoReturn for exhaustiveness checking <https://github.com/python/mypy/issues/5818>`_
+
+The Steering Council is not well-placed to solve the above problems
+-------------------------------------------------------------------
+
+The SC has the entire language in its remit, and is not well-placed to make
+decisions that are purely about the type system -- if only because they don't have
+the time to deal with type system arcana alongside their other responsibilities.
+This is similar in spirit to the reasons why the Steering Council sometimes uses
+PEP delegation.
+
+Specification
+=============
+
+We propose the creation of a new group, the Typing Council. This group will
+be responsible for developing and maintaining the Python type system, and
+for solving the above problems.
+
+The "operations and process" section describes how this group would operate and
+be governed.
+
+The more exciting "projects" section describes solutions to the above problems
+that the Typing Council could shepherd.
+
+Mandate
+-------
+
+The Typing Council's mandate is to ensure that the Python type system is:
+
+* **Useful**: The type system should serve common use cases. As identified
+  by :pep:`484`, the primary use case is static analysis, but there are others,
+  such as runtime type checking, static compilation, IDE support, and documentation.
+  The Typing Council should consider all of these use cases when making decisions,
+  and be open to supporting additional use cases as they come up.
+* **Usable**: The type system should be easy to use for Python developers. It
+  should be ergonomic to write well-typed Python code that is accepted by type
+  checkers. There should be good documentation for the type system.
+* **Stable**: As the type system matures, users should be able to rely on their
+  typed code continuing to work and be able to trust their mental model for the
+  type system. Changes should be made with care and in a way
+  that minimizes disruption. Nevertheless, the type system should be able to
+  evolve, and type checkers need not be held to the same strict compatibility
+  standards as the Python language itself.
```

Comment:
It might be worth expanding on this. Here's a few thoughts:

- The symbols in the typing module should be held to the same standards of compatibility as anything else in the standard library, i.e. we can't just make `typing.Something` disappear at any point, only after we've gone through DperecationWarnings for a few releases.
- But for most of the Council's decisions, the compatibility we're concerned about would be in the behavior of type checkers, and there I don't think it makes sense to use Python's backward compatibility policy. Type checkers make backward-incompatible changes all the time when they tweak their type inference, and they don't follow Python's versioning model.

Reply:
That sounds about right, with one exception: If there are gray areas where all type checkers currently agree, even if we come up with a proposal for better semantics, I think we ought to be pretty conservative. I believe mypy and pyright (at least) often solve this by keeping the default behavior backwards compatible but offer a flag (sadly type-checker-specific) to enable the "better" approach. For example, mypy started out without `Optional` checking (hard to believe now!) and took a very long time to change the default.

The reason to be conservative in such cases is that if the type checker changes, that often forces maintainers to update their code (if they have policies in place to require passing type checks, which many projects do). I guess you and Shantanu have plenty of experience with how to handle that -- how conservative you should be often depends on how much churn we expect in the community as a result of the change.

Reply:
I will push a rewrite of this last sentence, let's see if that works better.